sdk/go/common/resource/plugin/analyzer_plugin.go

```diff
@@ -140,8 +142,47 @@ func NewPolicyAnalyzer(
 		}
 	}
 
+	handshake := func(
+		ctx context.Context, bin string, prefix string, conn *grpc.ClientConn,
+	) (*pulumirpc.AnalyzerHandshakeResponse, error) {
+		// For analyzers the root directory and program directory are the location of the PulumiPolicy.yaml _not_ the
+		// location of the shim plugin.
+		dir := policyPackPath
+		client := pulumirpc.NewAnalyzerClient(conn)
+
+		mconfig, err := MarshalProperties(resource.ToResourcePropertyMap(opts.Config),
+			MarshalOptions{KeepSecrets: true})
+		if err != nil {
+			return nil, fmt.Errorf("marshalling config: %w", err)
+		}
+
+		res, err := client.Handshake(ctx, &pulumirpc.AnalyzerHandshakeRequest{
+			EngineAddress:    host.ServerAddr(),
+			RootDirectory:    &dir,
+			ProgramDirectory: &dir,
+
+			Stack:        opts.Stack,
+			Project:      opts.Project,
+			Organization: opts.Organization,
+
+			DryRun: opts.DryRun,
+			Config: mconfig,
+		})
+		if err != nil {
+			status, ok := status.FromError(err)
+			if ok && status.Code() == codes.Unimplemented {
+				// If the provider doesn't implement Handshake, that's fine -- we'll fall back to existing behavior.
+				logging.V(7).Infof("Handshake: not supported by '%v'", bin)
+				return nil, nil
+			}
+		}
+
+		logging.V(7).Infof("Handshake: success [%v]", bin)
+		return res, nil
+	}
+
 	plug, _, err := newPlugin(ctx, pwd, pluginPath, fmt.Sprintf("%v (analyzer)", name),
-		apitype.AnalyzerPlugin, args, env, testConnection,
+		apitype.AnalyzerPlugin, args, env, handshake,
```

Comment:
What's the difference between `NewAnalyzer` and `NewPolicyAnalyzer`, and should we be doing the handshake for in `NewAnalyzer` too?

Reply:
NewAnalyzer should just be removed, it's not used.